files/private-chef-ctl-commands/key-control.rb

```diff
@@ -0,0 +1,255 @@
+        row[:public_key].each_line do |line|
+          puts "        #{line}"
+        end
+      end
+    end
+  end
+end
+
+class ArgParser
+  def find_arg_after_identifier(args, identifier)
+    result = nil
+    args.each_with_index do |arg, i|
+      if arg == identifier
+        result = args[i+1]
+        break
+      end
+    end
+    result
+  end
+end
+
+class AddKey
+  def check_org_args(cmd_args, usage_msg)
+    unless cmd_args.length == 2 || \
+      (cmd_args.length == 4 && (cmd_args[2] == "--expiration-date" || cmd_args[2] == "--key-name")) || \
+      (cmd_args.length == 6 && ((cmd_args[2] == "--expiration-date" && cmd_args[4] == "--key-name") || (cmd_args[4] == "--expiration-date" && cmd_args[2] == "--key-name")))
+      STDERR.puts usage_msg
+      STDERR.puts "Usage: Expiration date defaults to infinite. Pass an ISO 8601 fomatted string: YYYY-MM-DDTHH:MM:SS e.g. 2013-12-24T21:00:00"
+      STDERR.puts "Usage: Default name used is the fingerprint of the key passed."
+      exit 1
+    end
+  end
+
+  def check_client_args(cmd_args, usage_msg)
+    unless cmd_args.length == 3 || \
+      (cmd_args.length == 5 && (cmd_args[3] == "--expiration-date" || cmd_args[3] == "--key-name")) || \
+      (cmd_args.length == 7 && ((cmd_args[3] == "--expiration-date" && cmd_args[5] == "--key-name") || (cmd_args[5] == "--expiration-date" && cmd_args[3] == "--key-name")))
+      STDERR.puts usage_msg
+      STDERR.puts "Usage: Expiration date defaults to infinite. Pass an ISO 8601 fomatted string: YYYY-MM-DDTHH:MM:SS e.g. 2013-12-24T21:00:00"
+      STDERR.puts "Usage: Default name used is the fingerprint of the key passed."
+      exit 1
+    end
+  end
+
+  def add_key(cmd_args, id)
+    # TODO should keyname be optional and default the key fingerprint?
+
+    # this is guaranteed to work because it would have crashed above if not
+    # expiration_date and key_name will be nil if they weren't passed
+    arg_parser = ArgParser.new
+    expiration_date = arg_parser.find_arg_after_identifier(cmd_args, "--expiration-date")
+    key_name        = arg_parser.find_arg_after_identifier(cmd_args, "--key-name")
+    user = cmd_args[0]
+    key = File.read(cmd_args[1])
+
+    unless /-----BEGIN PUBLIC KEY-----/.match(key)
+      STDERR.puts "Error: Invalid public key passed. Key must begin with:"
+      STDERR.puts "Error: -----BEGIN PUBLIC KEY-----"
+      exit 1
+    end
+
+    # if a key_name was not specified, use the fingerprint of the key
+    unless key_name
+      begin
+        openssl_key_object = OpenSSL::PKey::RSA.new(key)
+        data_string = OpenSSL::ASN1::Sequence([
+                                               OpenSSL::ASN1::Integer.new(openssl_key_object.public_key.n),
+                                               OpenSSL::ASN1::Integer.new(openssl_key_object.public_key.e)
+                                              ])
+        key_name = OpenSSL::Digest::SHA1.hexdigest(data_string.to_der).scan(/../).join(':')
```

Comment:
Alrighty, this looks good to me.  But, since it took me a while to understand, I figured leaving a note might be helpful for others:

1. We use PKCS#1 public and private key files.  In this case the user is providing a public key.

2. openssl_key_object.public_key.n extracts the RSA modulus, openssl_key_object.public_key.e extracts the public exponent.  These same values are also contained in the private key and thus we could easily write a tool for users to figure out which fingerprint matches the private key they have on disk.

3. OpenSSL::ASN1 is a way to represent data structures that can later be encoded for data interchange.

4. OpenSSL:ASN1::Sequence creates a new data structure from a list of OpenSSL:ASN1 base times (here, we have a sequence of integers.

5. data_string.to_der takes our sequence and encodes it using DER (Distinguished Encoding Rules).  This gives us a sequence of bytes that ruby represents as a string.

6. We then take SHA1 of this string as hex and add ":"'s for readability

As mentioned in (2), I think it would be nice to also provide a tool that shows the user the fingerprint of their private key (perhaps a knife command).